File: scripts/tools/github_adapter.py
#!/usr/bin/env python3
"""
GitHub Tool Adapter — read-only GitHub API interactions.

Read-only actions use real API calls when GITHUB_TOKEN is available.
Write actions remain scaffold-only (never auto-executed).
Falls back to scaffold data when no token is present.
"""

from __future__ import annotations

import json
import os
from typing import Any, Dict, FrozenSet, Optional
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

from tools.base_adapter import BaseToolAdapter, ToolAction, ToolResult

GITHUB_API = "https://api.github.com"
TIMEOUT_SECONDS = 15


class GitHubAdapter(BaseToolAdapter):
    """Adapter for GitHub API interactions."""

    READ_ACTIONS = frozenset({"read_pr", "read_issue", "list_files", "read_commit"})
    WRITE_ACTIONS = frozenset({"create_pr"})

    @property
    def name(self) -> str:
        return "github"

    @property
    def supported_actions(self) -> FrozenSet[str]:
        return self.READ_ACTIONS | self.WRITE_ACTIONS

    def check_auth(self) -> bool:
        return bool(os.environ.get("GITHUB_TOKEN"))

    @property
    def _token(self) -> Optional[str]:
        return os.environ.get("GITHUB_TOKEN")

    def _api_get(self, path: str) -> Dict[str, Any]:
        """Make an authenticated GET request to the GitHub API."""
        url = f"{GITHUB_API}/{path.lstrip('/')}"
        headers = {"Accept": "application/vnd.github+json"}
        if self._token:
            headers["Authorization"] = f"Bearer {self._token}"
        req = Request(url, headers=headers, method="GET")
        with urlopen(req, timeout=TIMEOUT_SECONDS) as resp:
            return json.loads(resp.read().decode())

    def execute_action(self, action: ToolAction) -> ToolResult:
        handler = {
            "read_pr": self._read_pr,
            "read_issue": self._read_issue,
            "list_files": self._list_files,
            "read_commit": self._read_commit,
            "create_pr": self._create_pr,
        }.get(action.action)

        if not handler:
            return ToolResult(
                tool_name=self.name, action=action.action,
                success=False, error=f"Unsupported action: {action.action}",
            )
        return handler(action)

    def _read_pr(self, action: ToolAction) -> ToolResult:
        owner = action.params.get("owner", "")
        repo = action.params.get("repo", "")
        number = action.params.get("number", "")
        if not all([owner, repo, number]) or not self._token:
            return self._scaffold("read_pr", action)
        try:
            data = self._api_get(f"repos/{owner}/{repo}/pulls/{number}")
            return ToolResult(tool_name=self.name, action="read_pr", success=True, data=data)
        except (HTTPError, URLError, TimeoutError) as e:
            return ToolResult(tool_name=self.name, action="read_pr", success=False, error=str(e))

    def _read_issue(self, action: ToolAction) -> ToolResult:
        owner = action.params.get("owner", "")
        repo = action.params.get("repo", "")
        number = action.params.get("number", "")
        if not all([owner, repo, number]) or not self._token:
            return self._scaffold("read_issue", action)
        try:
            data = self._api_get(f"repos/{owner}/{repo}/issues/{number}")
            return ToolResult(tool_name=self.name, action="read_issue", success=True, data=data)
        except (HTTPError, URLError, TimeoutError) as e:
            return ToolResult(tool_name=self.name, action="read_issue", success=False, error=str(e))

    def _list_files(self, action: ToolAction) -> ToolResult:
        owner = action.params.get("owner", "")
        repo = action.params.get("repo", "")
        number = action.params.get("number", "")
        if not all([owner, repo, number]) or not self._token:
            return self._scaffold("list_files", action)
        try:
            data = self._api_get(f"repos/{owner}/{repo}/pulls/{number}/files")
            return ToolResult(tool_name=self.name, action="list_files", success=True, data=data)
        except (HTTPError, URLError, TimeoutError) as e:
            return ToolResult(tool_name=self.name, action="list_files", success=False, error=str(e))

    def _read_commit(self, action: ToolAction) -> ToolResult:
        owner = action.params.get("owner", "")
        repo = action.params.get("repo", "")
        sha = action.params.get("sha", "")
        if not all([owner, repo, sha]) or not self._token:
            return self._scaffold("read_commit", action)
        try:
            data = self._api_get(f"repos/{owner}/{repo}/commits/{sha}")
            return ToolResult(tool_name=self.name, action="read_commit", success=True, data=data)
        except (HTTPError, URLError, TimeoutError) as e:
            return ToolResult(tool_name=self.name, action="read_commit", success=False, error=str(e))

    def _create_pr(self, action: ToolAction) -> ToolResult:
        """Write action — scaffold only, never auto-executed."""
        return self._scaffold("create_pr", action)

    def _scaffold(self, action_name: str, action: ToolAction) -> ToolResult:
        """Return scaffold data when no token or params are missing."""
        return ToolResult(
            tool_name=self.name, action=action_name, success=True,
            data={"scaffold": True, "action": action_name, "params": action.params},
        )

File: src/scripts/tools/github_adapter.ts
/**
 * GitHub Tool Adapter — read-only GitHub API interactions.
 *
 * Read-only actions use real API calls when GITHUB_TOKEN is available.
 * Write actions remain scaffold-only (never auto-executed).
 * Falls back to scaffold data when no token is present.
 *
 * TypeScript twin of `src/scripts/tools/github_adapter.py` (ADR-094 —
 * Python→TS migration, Phase 1 / tools adapter cluster). Public API mirrors
 * the Python module exactly (snake_case kept deliberately): the
 * `read_pr` / `read_issue` / `list_files` / `read_commit` / `create_pr`
 * dispatch, the GITHUB_TOKEN credential-from-env handling, the scaffold
 * fallback shape, and the GitHub API request construction (URL, headers,
 * method). The Python reference uses `urllib.request`; this twin builds the
 * identical request and performs the fetch through `_httpGetJson` — a seam
 * tests stub so no live network call is made (the no-token path scaffolds and
 * never reaches HTTP, exactly like the original). No behaviour changes.
 */
import { spawnSync } from 'node:child_process';

import { BaseToolAdapter, ToolAction, ToolResult } from './base_adapter.js';

export const GITHUB_API = 'https://api.github.com';
export const TIMEOUT_SECONDS = 15;

/** Request shape mirroring `urllib.request.Request`. */
export interface BuiltRequest {
    url: string;
    headers: Record<string, string>;
    method: string;
    timeout: number;
}

/** Adapter for GitHub API interactions. */
export class GitHubAdapter extends BaseToolAdapter {
    static readonly READ_ACTIONS: ReadonlySet<string> = new Set([
        'read_pr',
        'read_issue',
        'list_files',
        'read_commit',
    ]);
    static readonly WRITE_ACTIONS: ReadonlySet<string> = new Set(['create_pr']);

    get name(): string {
        return 'github';
    }

    get supported_actions(): ReadonlySet<string> {
        return new Set([...GitHubAdapter.READ_ACTIONS, ...GitHubAdapter.WRITE_ACTIONS]);
    }

    check_auth(): boolean {
        return Boolean(process.env.GITHUB_TOKEN);
    }

    protected get _token(): string | undefined {
        return process.env.GITHUB_TOKEN || undefined;
    }

    /**
     * Build the GitHub API request for `path` (mirrors the `urllib.Request`
     * construction in `_api_get`). Exposed for parity tests.
     */
    _buildRequest(path: string): BuiltRequest {
        const url = `${GITHUB_API}/${path.replace(/^\/+/, '')}`;
        const headers: Record<string, string> = { Accept: 'application/vnd.github+json' };
        if (this._token) {
            headers.Authorization = `Bearer ${this._token}`;
        }
        return { url, headers, method: 'GET', timeout: TIMEOUT_SECONDS };
    }

    /**
     * Perform the authenticated GET and decode JSON. Seam over the actual
     * network call — never reached on the no-token / scaffold path. Default
     * implementation mirrors `urlopen(...).read().decode()` + `json.loads`.
     */
    protected _httpGetJson(req: BuiltRequest): Record<string, unknown> {
        const headerArgs: string[] = [];
        for (const [k, v] of Object.entries(req.headers)) {
            headerArgs.push('-H', `${k}: ${v}`);
        }
        const proc = spawnSync(
            'curl',
            ['-sS', '--fail', '--max-time', String(req.timeout), ...headerArgs, req.url],
            { encoding: 'utf8' },
        );
        if (proc.status !== 0) {
            throw new Error(proc.stderr || `request failed for ${req.url}`);
        }
        return JSON.parse(proc.stdout) as Record<string, unknown>;
    }

    private _api_get(path: string): Record<string, unknown> {
        return this._httpGetJson(this._buildRequest(path));
    }

    execute_action(action: ToolAction): ToolResult {
        const handlers: Record<string, ((a: ToolAction) => ToolResult) | undefined> = {
            read_pr: (a) => this._read_pr(a),
            read_issue: (a) => this._read_issue(a),
            list_files: (a) => this._list_files(a),
            read_commit: (a) => this._read_commit(a),
            create_pr: (a) => this._create_pr(a),
        };
        const handler = handlers[action.action];

        if (!handler) {
            return new ToolResult({
                tool_name: this.name,
                action: action.action,
                success: false,
                error: `Unsupported action: ${action.action}`,
            });
        }
        return handler(action);
    }

    private _read_pr(action: ToolAction): ToolResult {
        const owner = (action.params.owner as string | undefined) ?? '';
        const repo = (action.params.repo as string | undefined) ?? '';
        const number = (action.params.number as string | undefined) ?? '';
        if (!(owner && repo && number) || !this._token) {
            return this._scaffold('read_pr', action);
        }
        try {
            const data = this._api_get(`repos/${owner}/${repo}/pulls/${number}`);
            return new ToolResult({ tool_name: this.name, action: 'read_pr', success: true, data });
        } catch (e) {
            return new ToolResult({
                tool_name: this.name,
                action: 'read_pr',
                success: false,
                error: String(e instanceof Error ? e.message : e),
            });
        }
    }

    private _read_issue(action: ToolAction): ToolResult {
        const owner = (action.params.owner as string | undefined) ?? '';
        const repo = (action.params.repo as string | undefined) ?? '';
        const number = (action.params.number as string | undefined) ?? '';
        if (!(owner && repo && number) || !this._token) {
            return this._scaffold('read_issue', action);
        }
        try {
            const data = this._api_get(`repos/${owner}/${repo}/issues/${number}`);
            return new ToolResult({
                tool_name: this.name,
                action: 'read_issue',
                success: true,
                data,
            });
        } catch (e) {
            return new ToolResult({
                tool_name: this.name,
                action: 'read_issue',
                success: false,
                error: String(e instanceof Error ? e.message : e),
            });
        }
    }

    private _list_files(action: ToolAction): ToolResult {
        const owner = (action.params.owner as string | undefined) ?? '';
        const repo = (action.params.repo as string | undefined) ?? '';
        const number = (action.params.number as string | undefined) ?? '';
        if (!(owner && repo && number) || !this._token) {
            return this._scaffold('list_files', action);
        }
        try {
            const data = this._api_get(`repos/${owner}/${repo}/pulls/${number}/files`);
            return new ToolResult({
                tool_name: this.name,
                action: 'list_files',
                success: true,
                data,
            });
        } catch (e) {
            return new ToolResult({
                tool_name: this.name,
                action: 'list_files',
                success: false,
                error: String(e instanceof Error ? e.message : e),
            });
        }
    }

    private _read_commit(action: ToolAction): ToolResult {
        const owner = (action.params.owner as string | undefined) ?? '';
        const repo = (action.params.repo as string | undefined) ?? '';
        const sha = (action.params.sha as string | undefined) ?? '';
        if (!(owner && repo && sha) || !this._token) {
            return this._scaffold('read_commit', action);
        }
        try {
            const data = this._api_get(`repos/${owner}/${repo}/commits/${sha}`);
            return new ToolResult({
                tool_name: this.name,
                action: 'read_commit',
                success: true,
                data,
            });
        } catch (e) {
            return new ToolResult({
                tool_name: this.name,
                action: 'read_commit',
                success: false,
                error: String(e instanceof Error ? e.message : e),
            });
        }
    }

    /** Write action — scaffold only, never auto-executed. */
    private _create_pr(action: ToolAction): ToolResult {
        return this._scaffold('create_pr', action);
    }

    /** Return scaffold data when no token or params are missing. */
    private _scaffold(action_name: string, action: ToolAction): ToolResult {
        return new ToolResult({
            tool_name: this.name,
            action: action_name,
            success: true,
            data: { scaffold: true, action: action_name, params: action.params },
        });
    }
}
